File: credit.c
#include <cs50.h>
#include <stdio.h>

int count_ccn(long a);
bool check_sum(int b, long c);

int main(void)
{
    long ccn;
    do
    {
        ccn = get_long("CCN: ");
    }
    while (ccn <= 0);

    int num = count_ccn(ccn);

    if (check_sum(num, ccn) == true)
    {
        // American Express uses 15-digit numbers & start with 34 or 37
        if ((ccn >= 34e13 && ccn < 35e13) || (ccn >= 37e13 && ccn < 38e13))
        {
            printf("AMEX\n");
        }

        // MasterCard uses 16-digit numbers & start with 51, 52, 53, 54, or 55
        else if (ccn >= 51e14 && ccn < 56e14)
        {
            printf("MASTERCARD\n");
        }

        // Visa uses 13- and 16-digit numbers & start with 4
        else if ((ccn >= 4e12 && ccn < 5e12) || (ccn >= 4e15 && ccn < 5e15))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

bool check_sum(int b, long c)
{
    // Guarda los impares, multiplicalos por dos y suma sus dígitos
    int cont = 0;
    for (long i = c, j = 10; i > 0; i /= 10, j *= 10)
    {
        int a = (((i % (j * 10)) - (i % j)) / j) * 2;
        int suma = (a / 10) + (a % 10);

        cont = cont + suma;
    }

    // Guarda los pares y sumamos los dígitos. Al final agregamos la suma anterior.
    int con = 0;
    for (long i = c, j = 1; i > 0; i = i / 10)
    {
        int a = ((i % (j * 10) - (i % j)) / j);

        con = con + a;
        j = j * 10;
    }

    if (((cont + con) % 10) == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

// Contamos el número de dígitos de la credit card
int count_ccn(long a)
{
    // Guardamos el contador en count
    int count;
    for (count = 0; a > 0; count++)
    {
        // Vamos eliminando las unidades finales por cada paso
        a = a / 10;
    }
    return count;
}
/*
Multiply every other digit by 2, starting with the number’s second-to-last digit, and then add those products’ digits together.
Add the sum to the sum of the digits that weren’t multiplied by 2.
If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0), the number is valid!
*/

File: credit.py
from cs50 import get_int


def main():
    # Prompt the user for a Credit Card Number
    while True:
        ccn = get_int("Credit Card Number: ")
        if ccn > 0:
            break
        
    # Transform CCN to String
    ccni = str(ccn)
    
    # Encuentra la longitud del CCN
    longitud = len(ccni)
        
    # Helper function check_sum runs the luhn's algorithm
    if check_sum(longitud, ccni):
        if (ccn >= 34e13 and ccn < 35e13) or (ccn >= 37e13 and ccn < 38e13):
            print("AMEX")
                
        elif (ccn >= 51e14 and ccn < 56e14):
            print("MASTERCARD")
            
        elif (ccn >= 4e12 and ccn < 5e12) or (ccn >= 4e15 and ccn < 5e15):
            print("VISA")
            
        else:
            print("INVALID")
                
    else:
        print("INVALID")
    
    
def check_sum(l, b):

    # Declare three variables to cero. The two sums and a counter
    pares = 0
    impares = 0
    
    for i in (b):
        if l % 2 == 0:
            a = int(i) * 2
            casted = str(a)
            for i in (casted):
                impares += int(i)
            l += 1
            
        else:
            pares += int(i)
            l -= 1

    if (pares + impares) % 10 == 0:
        return True
    else:
        return False


main()